### PR TITLE
Update for state handling

### PR DESCRIPTION
**Pull request by ujfalusi**
Hi,

This is a followup series for #3225 to update the fwe_state handling and to update the flow diagram.

Notable change is that we are going to try to recover from a BOOT_FAILED state via suspend/resume and that after a FW_READY_FAILED we move to BOOT_FAILED to make the state handling simpler.

The other bigger change is that we only allow IPC tx in BOOT_COMPLETED state

### CODE REVIEW

**Review by ranj063 — DISMISSED** [review]
looks good. Thanks @ujfalusi

**Review by plbossart — DISMISSED** [review]

File: sound/soc/sof/core.c [hunk 1]
```diff
@@ -202,6 +202,7 @@ static int sof_probe_continue(struct snd_sof_dev *sdev)
 	if (ret < 0) {
```
Comment (kv2019i) [inline sound/soc/sof/core.c:202]: s/ASOC/ASoC/ in commit tiel

File: sound/soc/sof/pm.c [hunk 2]
```diff
@@ -315,10 +315,11 @@ int snd_sof_prepare(struct device *dev)
...
 
 	/*
-	 * if the firmware is crashed then we try to aim for S3 to reboot the
-	 * firmware
+	 * if the firmware is crashed or boot failed then we try to aim for S3
+	 * to reboot the firmware
 	 */
-	if (sdev->fw_state == SOF_FW_CRASHED)
+	if (sdev->fw_state == SOF_FW_CRASHED ||
+	    sdev->fw_state == SOF_FW_BOOT_FAILED)
```
Comment (kv2019i) [inline sound/soc/sof/pm.c:322]: btw, @keyonjie , when we add the boot-from-IMR logic, we should probably have similar logic to stop using the IMR copy and do a cold boot if FW has crashed or boot failed. now the info is there, so this should be straightforward to add. Just a comment, not for this PR.

**Review by kv2019i — DISMISSED** [review]
One commit-typo, but nothing major.

File: sound/soc/sof/core.c (hunk 1 above)
Comment (ujfalusi) [reply sound/soc/sof/core.c:202]: oops

**Comment by ujfalusi** [comment]
Changes since v1:
- Fix subject of patch 4 (s/ASOC/ASoC)

**Review by plbossart — APPROVED** [review]

**Comment by plbossart** [comment]
@kv2019i @ranj063 @bardliao can you guys review this PR, it's 10+ days old.

**Review by bardliao — APPROVED** [review]
Sorry for the late. It looks good to me. Thanks.

**Review by ranj063 — APPROVED** [review]

**Pull request merged**